ui/app/models/totp.js

```diff
@@ -0,0 +1,173 @@
+  ],
+  period: [{ type: 'number', message: 'Period must be a number.' }],
+
+  skew: [{ validator: (value) => SKEW.includes(value), message: 'Skew must be one of ' + SKEW.join(', ') }],
+};
+
+@withModelValidations(validations)
+@withExpandedAttributes()
+export default class TotpModel extends Model {
+  @attr('string', {
+    readOnly: true,
+  })
+  backend;
+
+  @alias('account_name') name;
+
+  @attr('string', {
+    fieldValue: 'name',
+    editDisabled: true,
+  })
+  account_name;
+  @attr('string', {
+    editDisabled: true,
+    possibleValues: ALGORITHMS,
+    defaultValue: 'SHA1',
+  })
+  algorithm;
+  @attr('number', {
+    editDisabled: true,
+    possibleValues: DIGITS,
+    defaultValue: 6,
+  })
+  digits;
+  @attr('string', {
+    editDisabled: true,
+  })
+  issuer;
+  @attr('number', {
+    editDisabled: true,
+    defaultValue: 30,
+  })
+  period;
+
+  @attr('boolean', {
+    defaultValue: false,
+    label: 'Use Vault as provider for this key',
+    editDisabled: true,
+  })
+  generate;
+
+  // Used when generate is true
+  @attr('number', {
+    //defaultValue: 20,
+    editDisabled: true,
+  })
+  key_size;
+  @attr('number', {
+    possibleValues: SKEW,
+    //defaultValue: 1,
+    editDisabled: true,
+  })
+  skew;
+  @attr('boolean', {
+    //defaultValue: true,
+    editDisabled: true,
+  })
+  exported;
+
+  // Doesn't really make sense as we can generate our own QR code from the url
+  @attr('number', {
+    defaultValue: 0,
+    editDisabled: true,
+  })
+  qr_size;
+
+  // Used when generate is false
+  @attr('string', {
+    editDisabled: true,
+    label: 'otpauth url',
+    helpText: 'If a URL is provided the other fields can be left empty. E.g. otpauth://totp/Vault:test@test.com?secret=Y64VEVMBTSXCYIWRSHRNDZW62MPGVU2G&issuer=Vault',
+  })
+  url;
+  @attr('string', {
+    editDisabled: true,
+    label: 'Shared master key',
+  })
+  key;
+
+  // Returned when a key is created as provider
+  @attr('string', {
+    readOnly: true,
+  })
+  barcode;
+
+  get attrs() {
+    const keys = ['account_name', 'algorithm', 'digits', 'issuer', 'period'];
+    return keys.map((k) => this.allByKey[k]);
+  }
+
+  get generatedAttrs() {
+    const keys = ['url'];
+    return keys.map((k) => this.allByKey[k]);
+  }
+
+  @computed('generate', function () {
+    const defaultFields = ['generate'];
+    const options = ['algorithm', 'digits', 'period'];
+    const providerOptions = [];
+
+    if (this.generate) {
+      providerOptions.push('key_size', 'skew', 'exported', 'qr_size');
+    } else {
+      defaultFields.push('url', 'key');
+    }
+
+    defaultFields.push('account_name', 'issuer');
+
+    const groups = [
+      { default: defaultFields },
+      {
+        Options: [...options],
+      },
+    ];
+
+    if (this.generate) {
+      groups.push({
+        'Provider options': [...providerOptions],
+      });
+    }
+
+    return this._expandGroups(groups);
+  })
+  fieldGroups;
```

Comment:
The newer pattern is to keep models "thin" and any form logic should be handled by the form component, not the model. Instead, we can use the `@withFormFields` decorator which will "expand" all of the model attributes into the form structure we want (with the options object that houses editType and all of that info). 

The decorator can be called with args, like this: https://github.com/hashicorp/vault/blob/main/ui/app/models/pki/issuer.js#L27